alea/include/alps/alea.hpp

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 1998-2017 ALPS Collaboration. See COPYRIGHT.TXT
+ * All rights reserved. Use is subject to license terms. See LICENSE.TXT
+ * For use in publications, see ACKNOWLEDGE.TXT
+ */
+#pragma once
+
+/**
+ * @namespace alps::alea
+ * @brief Set of accumulators and statistical pre-/post-processing operations.
+ *
+ * Overview
+ * --------
+ * TODO
+ *
+ * Accumulators and results
+ * ------------------------
+ * Most accumulators (`mean_acc`) have a matching result class (`mean_result`).
+ * To obtain a result from an accumulator, the accumulators provide both a
+ * `result()` and a `finalize()` method, where
+ *
+ *  1. the `result()` method creates an intermediate result, which leaves the
+ *     accumulator untouched and thus must involve a copy of the data, while
+ *
+ *  2. the `finalize()` method invalidates the accumulator and thus allows to
+ *     repurpose its data as the simulation result.  The reset method then
+ *     re-creates an empty accumulator with the same size.
+ *
+ * This can be represented by the following finite state machine:
+ *
+ *                     c'tor   _______________      _______________
+ *                    ------->|               |    |               |  default
+ *     result, <<        <<   |     empty     |    | uninitialized |   c'tor
+ *      +-------+       +-----|_______________|    |_______________|<<-------
+ *      |       |       |            | |
+ *      |     __V_______V____  reset | | reset  ________________
+ *      |    |               |--->---+ +---<---|                |
+ *      +----|  accumulating |                 |     invalid    |
+ *           |_______________|---------------->|________________|
+ *                                finalize
+ */
```

Comment:
Since its brought up by both @krivenko and @iskakoff:

There is a conceptual difference between `uninitialized` and `invalid` state (I tried to picture this in the FSM diagram in the code linked to this comment.)

 * An *uninitialized* accumulator is one that has been default-constructed and is thus not in a usable state. (You can not call any method on it, including `reset()`.)

 * An *invalid* accumulator is one on which `finalize()` has been called and thus has given its data away to a result.  However, you can bring it back to a usable state by calling `reset()` on it.

I introduced the *uninitialized* state in response to @kleinhenz's comment.  I want accumulators to be default-constructible; however, what we can discuss whether it would be better if we have a sensible default for the size: either `0` as for Eigen vectors or `1` to make the scalar case the default.

Secondly, I agree that it is hard to distinguish `uninitialized_accumulator` and `invalid_accumulator`. I propose to rename `invalid_accumulator` to `finalized_accumulator`, then it is clear what happened to it.

Reply:
> I introduced the uninitialized state in response to @kleinhenz's comment. I want accumulators to be default-constructible; however, what we can discuss whether it would be better if we instead scrap the uninitialized state, and have a sensible default for the size: either 0 as for Eigen vectors or 1 to make the scalar case the default.

I think `size == 1` is a reasonable default in many cases. And it's certainly better than having the error-prone uninitialized state.